lib/charms/nginx_ingress_integrator/v0/nginx_route.py

```diff
@@ -0,0 +1,383 @@
+    owasp_modsecurity_custom_rules: typing.Optional[str] = None,
+    path_routes: typing.Optional[str] = None,
+    retry_errors: typing.Optional[str] = None,
+    rewrite_target: typing.Optional[str] = None,
+    rewrite_enabled: typing.Optional[bool] = None,
+    service_namespace: typing.Optional[str] = None,
+    session_cookie_max_age: typing.Optional[int] = None,
+    tls_secret_name: typing.Optional[str] = None,
+    nginx_route_relation_name: str = "nginx-route",
+) -> None:
+    """Set up nginx-route relation handlers on the requirer side.
+
+    This function must be invoked in the charm class constructor.
+
+    Args:
+        charm: The charm that requires the nginx-route relation.
+        service_hostname: configure Nginx ingress integrator
+            service-hostname option via relation.
+        service_name: configure Nginx ingress integrator service-name
+            option via relation.
+        service_port: configure Nginx ingress integrator service-port
+            option via relation.
+        additional_hostnames: configure Nginx ingress integrator
+            additional-hostnames option via relation, optional.
+        limit_rps: configure Nginx ingress integrator limit-rps
+            option via relation, optional.
+        limit_whitelist: configure Nginx ingress integrator
+            limit-whitelist option via relation, optional.
+        max_body_size: configure Nginx ingress integrator
+            max-body-size option via relation, optional.
+        owasp_modsecurity_crs: configure Nginx ingress integrator
+            owasp-modsecurity-crs  option via relation, optional.
+        owasp_modsecurity_custom_rules: configure Nginx ingress
+            integrator owasp-modsecurity-custom-rules option via
+            relation, optional.
+        path_routes: configure Nginx ingress integrator path-routes
+            option via relation, optional.
+        retry_errors: configure Nginx ingress integrator retry-errors
+            option via relation, optional.
+        rewrite_target: configure Nginx ingress integrator
+            rewrite-target option via relation, optional.
+        rewrite_enabled: configure Nginx ingress integrator
+            rewrite-enabled option via relation, optional.
+        service_namespace: configure Nginx ingress integrator
+            service-namespace option via relation, optional.
+        session_cookie_max_age: configure Nginx ingress integrator
+            session-cookie-max-age option via relation, optional.
+        tls_secret_name: configure Nginx ingress integrator
+            tls-secret-name option via relation, optional.
+        nginx_route_relation_name: Specifies the relation name of
+            the relation handled by this requirer class. The relation
+            must have the nginx-route interface.
+    """
+    config: typing.Dict[str, typing.Union[str, int, bool]] = {}
+    if service_hostname is not None:
```

Comment:
I do like the docstring and that the arguments are listed in the function. I can think of a few hacks to capture the kwargs whilst also retaining this, (e.g., involving decorators and functools.wraps), they aren't pretty tough.

One nice alternative that we could consider is uding a TypedDict because that also comes with autocomplete. Another option is to use a bunch of `typing.NamedTuple`, including aggregating similar arguments into sub tuples. That is something like what I have seen other libraries do in these cases. So you would have something like:

```python
# library code
from typing import NamedTuple

class Service(NamedTuple):
    hostname: str
    name: str
    ...

class Config(NamedTuple):
    service: Service
    ...

def require_nginx_route(..., config: Config, ...) -> None: ...

# user code
from <wherever> import Service, Config

class Charm(CharmBase):
    def __init__(self):
        service = Service(...)
        config = COnfig(service=service, ...)
        require_nginx_route(self, config, ...)
```

Reply:
The reason why I use function arguments here instead of `kwargs` is that:
- function arguments have better IDE support compare to `kwargs` with TypedDict
- function arguments are checked by Python, while the the type hint of `kwargs` is not enforced during the runtime
- TypedDict's support for omitting key requires Python 3.11

One concern about the NamedTuple approach is that users now require to import another class to use the charm library.

I think in this case the complexity of implementation (multiple `if`s) is acceptable for the simplicity of the interface.

Reply:
I think you are right that on balance, the arguments are a better way of doing things in this case

Just a note, for `TypedDict`, it has been possible I believe since their introduction to declare that not all keys have been required using the `total` argument. I have written a blog post on this topic: https://jdkandersson.com/2020/01/27/python-typeddict-arbitrary-key-names-with-totality/

Reply:
The problem for the Nginx ingress integrator charm configuration dictionary is that the `service-hostname`, `service-name`, and `service-port` are required, while other fields are not. If using the `total` argument in the `TypedDict`, all fields will be optional.

Reply:
It is possible to use a class hierarchy where some have the `total=True` and others `total=False`, the linked blog post above gives an example

Reply:
Learned, very nice solution 👍